src/main.cpp

```diff
@@ -1,82 +1,92 @@
-#include "STDIndicesStream.h"
-#elif defined(STD_RANGES)
-#include "STDRangesStream.hpp"
-#elif defined(TBB)
-#include "TBBStream.hpp"
-#elif defined(THRUST)
-#include "ThrustStream.h"
-#elif defined(HIP)
-#include "HIPStream.h"
-#elif defined(HC)
-#include "HCStream.h"
-#elif defined(OCL)
-#include "OCLStream.h"
-#elif defined(USE_RAJA)
-#include "RAJAStream.hpp"
-#elif defined(KOKKOS)
-#include "KokkosStream.hpp"
-#elif defined(ACC)
-#include "ACCStream.h"
-#elif defined(SYCL)
-#include "SYCLStream.h"
-#elif defined(SYCL2020)
-#include "SYCLStream2020.h"
-#elif defined(OMP)
-#include "OMPStream.h"
-#elif defined(FUTHARK)
-#include "FutharkStream.h"
-#endif
+using namespace std;
 
 // Default size of 2^25
 int ARRAY_SIZE = 33554432;
 unsigned int num_times = 100;
 unsigned int deviceIndex = 0;
 bool use_float = false;
 bool output_as_csv = false;
-bool mibibytes = false;
-std::string csv_separator = ",";
-
-template <typename T>
-void check_solution(const unsigned int ntimes, std::vector<T>& a, std::vector<T>& b, std::vector<T>& c, T& sum);
+Unit unit = MegaByte;
+string csv_separator = ",";
 
-template <typename T>
-void run();
+// Benchmarks:
+constexpr size_t num_benchmarks = 6;
+array<char const*, num_benchmarks> labels = {"Copy", "Add", "Mul", "Triad", "Dot", "Nstream"};
+// Weights data moved by benchmark & therefore achieved BW:
+// bytes = weight * sizeof(T) * ARRAY_SIZE -> bw = bytes / dur
+array<size_t, num_benchmarks> weight = {/*Copy:*/ 2, /*Add:*/ 2, /*Mul:*/ 3, /*Triad:*/ 3, /*Dot:*/ 2, /*Nstream:*/ 4};
 
 // Options for running the benchmark:
-// - All 5 kernels (Copy, Add, Mul, Triad, Dot).
-// - Triad only.
-// - Nstream only.
-enum class Benchmark {All, Triad, Nstream};
+// - Classic 5 kernels (Copy, Add, Mul, Triad, Dot).
+// - All kernels (Copy, Add, Mul, Triad, Dot, Nstream).
+// - Individual kernels only.
+enum class Benchmark : int {Copy = 0, Add = 1, Mul = 2, Triad = 3, Dot = 4, Nstream = 5, Classic, All};
 
 // Selected run options.
 Benchmark selection = Benchmark::All;
 
+// Returns true if the benchmark needs to be run:
+bool run_benchmark(int id) {
+  if (selection == Benchmark::All)                return true;
+  if (selection == Benchmark::Classic && id < 5)  return true;
+  if (id == 0 && selection == Benchmark::Copy)    return true;
+  if (id == 1 && selection == Benchmark::Add)     return true;
+  if (id == 2 && selection == Benchmark::Mul)     return true;
+  if (id == 3 && selection == Benchmark::Triad)   return true;
+  if (id == 4 && selection == Benchmark::Dot)     return true;
+  if (id == 5 && selection == Benchmark::Nstream) return true;
+  return false;
+}
+
+// Prints all available benchmark labels:
+template <typename OStream>
+void print_labels(OStream& os) {
+  for (int i = 0; i < num_benchmarks; ++i) {
+    os << labels[i];
+    if (i != (num_benchmarks - 1)) os << ",";
+  }
+}
+
+// Clock and duration types:
+using clk_t = chrono::high_resolution_clock;
+using dur_t = chrono::duration<double>;
```

Comment:
These two are not used that frequently in the new structure so for clarity just use without alias in `time`?